docs/fixing-warnings.md

```diff
@@ -0,0 +1,189 @@
+}
+```
+
+There aren't many workarounds for `RequiresUnreferencedCode`. The best way is to avoid calling
+the method at all when trimming and use something else which is trim-compatible. If you're
+writing a library and it's not in your control whether or not to call the method and you just
+want to communicate to *your* caller, you can also add `RequiresUnreferencedCode` to your own
+method. This silences all trimming warnings in your code, but will produce a warning whenever
+someone calls your method.
+
+If you can somehow determine that the call is safe, and all the code that's needed won't be
+trimmed away, you can also suppress the warning using
+[UnconditionalSuppressMessageAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.unconditionalsuppressmessageattribute?view=net-5.0).
+For example:
+
+```C#
+[RequiresUnreferencedCode("Use 'MethodFriendlyToTrimming' instead")]
+void MethodWithAssemblyLoad() { ... }
+
+[UnconditionalSuppressMessage("AssemblyLoadTrimming", "IL2026:RequiresUnreferencedCode",
+    Justification = "Everything referenced in the loaded assembly is manually preserved, so it's safe")]
+void TestMethod()
+{
+    MethodWithAssemblyLoad(); // Warning suppressed
+}
+```
+
+`UnconditionalSuppressMessage` is like `SuppressMessage` but it is preserved into IL, so the
+trimmer can see the suppression even after build and publish. `SuppressMessage` and `#pragma`
+directives are only present in source so they can't be used to silence warnings from the
+trimmer. Be very careful when suppressing trim warnings: it's possible that the call may be
+trim-compatible now, but as you change your code that may change and you may forget to review all
+the suppressions.
+
+### DynamicallyAccessedMembers
+
+`DynamicallyAccessedMembers` is usually about reflection. Unlike `RequiresUnreferencedCode`,
+reflection can sometimes be understood by the trimmer as long as it's annotated correctly.
+Let's take another look at the original example:
+
+```C#
+string s = Console.ReadLine();
+Type type = Type.GetType(s);
+foreach (var m in type.GetMethods())
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+In the example above, the real problem is `Console.ReadLine()`. Because *any* type could
+be read the trimmer has no way to know if you need methods on `System.Tuple` or `System.Guid`
+or any other type. On the other hand, if your code looked like,
+
+```C#
+Type type = typeof(System.Tuple);
+foreach (var m in type.GetMethods())
+{
+    Console.WriteLine(m.Name);
+}
+```
+
+This would be fine. Here the trimmer can see the exact type being referenced: `System.Tuple`. Now
+it can use flow analysis to determine that it needs to keep all public methods. So where does
+`DynamicallyAccessMembers` come in? What happens if the reflection is split across methods?
+
+```C#
+void M1()
+{
+    M2(typeof(System.Tuple));
+}
+void M2(Type type)
+{
+    // Trim analysis warning IL2070: net6.Program.M2(Type): 'this' argument does not satisfy
```

Comment:
Move out the error into a block to show what the compiler would say

Reply:
Sorry, not sure what you mean by this. You mean extract the comment into a separate code block?

Reply:
Yes, and you can mark it as console, something like:

-----

```c#
void M1()
{
    M2(typeof(System.Tuple));
}
void M2(Type type)
{
    var methods = type.GetMethods();
    ...
}
```

If you compile, you'll see the following warning:

```console
Trim analysis warning IL2070: net6.Program.M2(Type): 'this' argument does not satisfy
'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'System.Type.GetMethods()'. The
parameter 'type' of method 'net6.Program.M2(Type)' does not have matching annotations. The
source value must declare at least the same requirements as those declared on the target
location it is assigned to.
```